### PR TITLE
Staking: store last `min-active-bond` on-chain

### PR DESCRIPTION
**Pull request by gpestana**
Storing the `min-active-bond` onchain helps the UIs with minimal on-chain costs.

Closes https://github.com/paritytech/substrate/issues/12746

### CODE REVIEW

File: frame/staking/src/pallet/impls.rs [hunk 1]
```diff
@@ -793,6 +795,12 @@ impl<T: Config> Pallet<T> {
 			nominators_taken
 		);
 
+		// `all_voters` is sorted by vote weight, so we use the last element to set the current
+		// minimum active bond. Note that depending on the `T::VoterList` list, the sorting may be
+		// inconsistent as the sorting may be lazy.
+		let min_active_bond: T::CurrencyBalance = all_voters.last().map_or(0, |v| v.1).into();
```
Comment (gpestana) [inline frame/staking/src/pallet/impls.rs:801]: I wonder if we should iterate over `all_voters` and pick the min bond to account for inconsistencies in lazily sorted `T::VoterList` like bags-list.

Comment (Ank4n) [reply frame/staking/src/pallet/impls.rs:801]: We are already iterating over all_voters right? Is there a reason not to keep track of the minimum active bond in that loop?

Comment (Ank4n) [reply frame/staking/src/pallet/impls.rs:801]: We are iterating over `sorted_voters` to be accurate.

File: frame/staking/src/pallet/impls.rs [hunk 2]
```diff
@@ -727,6 +729,7 @@ impl<T: Config> Pallet<T> {
 		let mut nominators_taken = 0u32;
 
 		let mut sorted_voters = T::VoterList::iter();
+		let mut min_active_bond = None;
```
Comment (kianenigma) [inline frame/staking/src/pallet/impls.rs:732]: I have not seen (or like) using option like this. Start with `Bounded::max_value()` and go downward.

File: frame/staking/src/pallet/impls.rs [hunk 3]
```diff
@@ -785,6 +797,10 @@ impl<T: Config> Pallet<T> {
 			slashing_spans.len() as u32,
 		));
 
+		LastMinimumActiveBond::<T>::put::<T::CurrencyBalance>(
+			min_active_bond.map_or(0, |v| v).into(),
+		);
```
Comment (kianenigma) [inline frame/staking/src/pallet/impls.rs:802]:
```suggestion
		LastMinimumActiveBond::<T>::put(min_active_bond);
```

File: frame/staking/src/tests.rs [hunk 4]
```diff
@@ -4398,6 +4398,24 @@ mod election_data_provider {
...
 
+	#[test]
+	fn set_last_minimum_nominator_bond_is_correct() {
+		ExtBuilder::default()
+			.nominate(false)
+			.add_staker(61, 60, 2_000, StakerStatus::<AccountId>::Nominator(vec![21]))
+			.add_staker(71, 70, 10, StakerStatus::<AccountId>::Nominator(vec![21]))
+			.add_staker(81, 80, 50, StakerStatus::<AccountId>::Nominator(vec![21]))
+			.build_and_execute(|| {
+				assert_ok!(<Staking as ElectionDataProvider>::electing_voters(None));
```
Comment (kianenigma) [inline frame/staking/src/tests.rs:4409]: Can also test the initial condition.

**Comment by gpestana** [comment]
> if so, i think some names can reflect this a bit better.

I agree, I'm not fully happy with the naming `LastMinimumActiveBond` either. Renaming it to `MinimumActiveStake`.

File: frame/staking/src/pallet/impls.rs [hunk 5]
```diff
@@ -719,6 +722,7 @@ impl<T: Config> Pallet<T> {
 		let mut voters_seen = 0u32;
 		let mut validators_taken = 0u32;
 		let mut nominators_taken = 0u32;
+		let mut min_active_stake = u64::MAX;
```
Comment (Ank4n) [inline frame/staking/src/pallet/impls.rs:725]: I would have preferred to give it a type Option<u64> and initialize it to `None`.

Comment (Ank4n) [reply frame/staking/src/pallet/impls.rs:725]: Never mind, saw Kian's comment :joy:

**Review by Ank4n — APPROVED** [review]

**Comment by gpestana** [comment]
bot merge

**Pull request merged**